path/dynamic/dstarlite.go

```diff
@@ -0,0 +1,462 @@
+						}
+					}
+				}
+				d.update(s)
+			}
+		}
+	}
+}
+
+// Step performs one movement step along the best path towards the goal.
+// It returns false if no further progression toward the goal can be
+// achieved, either because the goal has been reached or because there
+// is no path.
+func (d *DStarLite) Step() bool {
+	/*
+	   procedure Main()
+	   {32”} while (s_start != s_goal)
+	   {33”} // if (rhs(s_start) = ∞) then there is no known path
+	   {34”}   s_start = argmin s'∈Succ(s_start)(c(s_start, s') + g(s'));
+	*/
+	if d.s.ID() == d.t.ID() {
+		return false
+	}
+	if math.IsInf(d.s.rhs, 1) {
+		return false
+	}
+	min := math.Inf(1)
+	var next *dStarLiteNode
+	for _, _s := range d.world.From(d.s) {
+		s := _s.(*dStarLiteNode)
+		w := d.weight(d.world.Edge(d.s, s)) + s.g
+		if w < min {
+			next = s
+			min = w
+		}
+	}
+	d.s = next
+
+	/*
+	   procedure Main()
+	   {35”}   Move to s_start;
+	*/
+	return true
+}
+
+// MoveTo moves to n in the world graph.
+func (d *DStarLite) MoveTo(n graph.Node) {
+	d.last = d.s
+	d.s = d.world.Node(n.ID()).(*dStarLiteNode)
+	d.keyModifier += d.heuristic(d.last, d.s)
+}
+
+// UpdateWorld updates or adds edges in the world graph. UpdateWorld will
+// panic if changes include a negative edge weight.
+func (d *DStarLite) UpdateWorld(changes []concrete.WeightedEdge) {
+	/*
+	   procedure Main()
+	   {36”}   Scan graph for changed edge costs;
+	   {37”}   if any edge costs changed
+	   {38”}     k_m = k_m + h(s_last, s_start);
+	   {39”}     s_last = s_start;
+	   {40”}     for all directed edges (u, v) with changed edge costs
+	   {41”}       c_old = c(u, v);
+	   {42”}       Update the edge cost c(u, v);
+	   {43”}       if (c_old > c(u, v))
+	   {44”}         if (u != s_goal) rhs(u) = min(rhs(u), c(u, v) + g(v));
+	   {45”}       else if (rhs(u) = c_old + g(v))
+	   {46”}         if (u != s_goal) rhs(u) = min s'∈Succ(u)(c(u, s') + g(s'));
+	   {47”}       UpdateVertex(u);
+	   {48”}     ComputeShortestPath()
+	*/
+	if len(changes) == 0 {
+		return
+	}
+	d.keyModifier += d.heuristic(d.last, d.s)
+	d.last = d.s
+	for _, e := range changes {
+		if e.Cost < 0 {
+			panic("D* Lite: negative edge weight")
+		}
+		cOld := d.weight(e.Edge)
+		u := d.worldNodeFor(e.From())
+		v := d.worldNodeFor(e.To())
+		d.world.SetEdge(concrete.Edge{F: u, T: v}, e.Cost)
+		if cOld > e.Cost {
+			if u.ID() != d.t.ID() {
+				u.rhs = math.Min(u.rhs, e.Cost+v.g)
+			}
+		} else if u.rhs == cOld+v.g {
+			if u.ID() != d.t.ID() {
+				u.rhs = math.Inf(1)
+				for _, t := range d.world.From(u) {
+					u.rhs = math.Min(u.rhs, d.weight(d.world.Edge(u, t))+t.(*dStarLiteNode).g)
+				}
+			}
+		}
+		d.update(u)
+	}
+	d.findShortestPath()
+}
+
+func (d *DStarLite) worldNodeFor(n graph.Node) *dStarLiteNode {
+	switch w := d.world.Node(n.ID()).(type) {
+	case *dStarLiteNode:
+		return w
+	case graph.Node:
+		panic(fmt.Sprintf("D* Lite: illegal world node type: %T", w))
```

Comment:
The cases that we have here are:

1. n exists in the world, so return it as w.

2. n doesn't exist in the world and d.world.Node(n.ID()) returns nil, so make a new node and return it (the default path).

3. There is a poisonous node in the world. Here we panic. This will only happen if the world handed to NewDStarLite already had been used and had a graph.Node that is not a *dStarLiteNode.